File: app/rotas/importacao.py
"""/importar - carga da planilha e do JSON do Trello, com pre-visualizacao."""

from __future__ import annotations

import csv
import io
from pathlib import Path

from fastapi import APIRouter, File, Form, Request, UploadFile
from fastapi.responses import Response
from sqlalchemy import select

from app.config import obter_config
from app.dependencias import SessaoDep, UsuarioDep
from app.modelos import MigracaoRejeitada, StgTrelloCartao
from app.servicos import (
    importacao_planilha,
    importacao_trello,
    reconciliacao as servico_reconciliacao,
)
from app.servicos.numeracao import recontar_sequencia
from app.web import marcar_download, pagina

rotas = APIRouter(tags=["importacao"])


@rotas.get("/importar")
def tela(request: Request, s: SessaoDep, usuario: UsuarioDep):
    usuario.exigir("catalogo.gerenciar")
    cfg = obter_config()
    return pagina(
        request,
        "paginas/importar.html",
        usuario=usuario,
        entrada=[
            p.name
            for p in sorted(cfg.caminho(cfg.dir_entrada).glob("*"))
            if p.suffix.lower() in (".xlsx", ".json")
        ],
        rejeitadas=list(
            s.execute(select(MigracaoRejeitada).order_by(MigracaoRejeitada.id.desc()).limit(100)).scalars()
        ),
        candidatos=list(
            s.execute(
                select(StgTrelloCartao).where(StgTrelloCartao.parecer_candidato.isnot(None))
            ).scalars()
        ),
    )


@rotas.get("/importar/reconciliacao")
def reconciliacao(request: Request, s: SessaoDep, usuario: UsuarioDep):
    """Os cinco relatórios da Fase 4 — expõem, nunca resolvem sozinhos."""
    usuario.exigir("catalogo.gerenciar")
    resultado = servico_reconciliacao.reconciliar(s)
    return pagina(
        request,
        "paginas/reconciliacao.html",
        usuario=usuario,
        reconciliacao=resultado,
        rejeitadas=servico_reconciliacao.rejeitadas(s),
    )


@rotas.get("/importar/reconciliacao.csv")
def reconciliacao_csv(request: Request, s: SessaoDep, usuario: UsuarioDep):
    usuario.exigir("exportar")
    resultado = servico_reconciliacao.reconciliar(s)
    buffer = io.StringIO(newline="")
    escritor = csv.writer(buffer, delimiter=";")
    escritor.writerow(["Relatório", "Referência", "Detalhe"])
    for titulo, _descricao, achados in resultado.como_secoes():
        for achado in achados:
            escritor.writerow([titulo, achado.referencia, achado.detalhe])
    return marcar_download(
        Response(
            b"\xef\xbb\xbf" + buffer.getvalue().encode("utf-8"),
            media_type="text/csv; charset=utf-8",
            headers={"Content-Disposition": 'attachment; filename="reconciliacao.csv"'},
        ),
        request,
    )


@rotas.post("/importar/planilha")
async def importar_planilha(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    arquivo: UploadFile = File(...),
    aplicar: str = Form(""),
):
    usuario.exigir("catalogo.gerenciar")
    cfg = obter_config()
    destino = cfg.caminho(cfg.dir_entrada) / (arquivo.filename or "planilha.xlsx")
    destino.write_bytes(await arquivo.read())

    relatorio = importacao_planilha.importar(s, destino, usuario, aplicar == "1")
    if aplicar == "1":
        recontar_sequencia(s)
        s.commit()
    else:
        s.rollback()
    return pagina(
        request,
        "paginas/importar_relatorio.html",
        usuario=usuario,
        relatorio=relatorio,
        aplicado=aplicar == "1",
        origem="planilha",
    )


@rotas.post("/importar/trello")
async def importar_trello(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    arquivo: UploadFile = File(...),
    aplicar: str = Form(""),
    baixar_anexos: str = Form(""),
):
    usuario.exigir("catalogo.gerenciar")
    cfg = obter_config()
    destino = cfg.caminho(cfg.dir_entrada) / (arquivo.filename or "trello.json")
    destino.write_bytes(await arquivo.read())

    relatorio = importacao_trello.importar(
        s,
        Path(destino),
        usuario,
        aplicar == "1",
        # sem marcar, nada e baixado: os anexos entram no relatorio de perdidos
        buscar_anexo=importacao_trello.baixar_url if baixar_anexos == "1" else None,
    )
    if aplicar == "1":
        s.commit()
    else:
        s.rollback()
    return pagina(
        request,
        "paginas/importar_relatorio_trello.html",
        usuario=usuario,
        relatorio=relatorio,
        aplicado=aplicar == "1",
    )

File: nuvem/src/rotas/importacao.ts
/**
 * /importar — carga da planilha e do JSON do Trello, com pré-visualização.
 * Porte de `app/rotas/importacao.py`.
 *
 * **Desvio da nuvem.** No Python o arquivo enviado era gravado em `entrada/`
 * (disco) e lido de lá; a tela listava o que havia na pasta. A função do
 * Netlify não tem disco: o arquivo é processado da memória e uma cópia vai para
 * o armazenamento (`entrada/<nome>`, a mesma guarda que a pasta fazia). A lista
 * "arquivos em entrada/" deixa de existir na tela — o armazenamento não é
 * pasta que se liste a cada abertura. Ver DESVIOS.md.
 */
import { Hono } from "hono";
import { desc, isNotNull } from "drizzle-orm";
import type { Ambiente, Ctx } from "../nucleo/contexto.js";
import { desfazer } from "../nucleo/contexto.js";
import { ErroHttp } from "../nucleo/erros.js";
import { formulario, usuarioCom } from "../dependencias.js";
import { marcarDownload, pagina } from "../web.js";
import { migracao_rejeitada, stg_trello_cartao } from "../db/esquema/index.js";
import { obterArmazenamento } from "../servicos/armazenamento.js";
import * as importacao_planilha from "../servicos/importacao_planilha.js";
import * as importacao_trello from "../servicos/importacao_trello.js";
import * as servico_reconciliacao from "../servicos/reconciliacao.js";
import { recontar_sequencia } from "../servicos/numeracao.js";
import { linha_csv } from "./relatorios.js";

export const rotas = new Hono<Ambiente>();

rotas.get("/importar", async (c) => {
  const usuario = await usuarioCom(c, "catalogo.gerenciar");
  const tx = c.get("tx");
  return pagina(c, "paginas/importar.html", usuario, {
    entrada: [],
    rejeitadas: await tx.select().from(migracao_rejeitada).orderBy(desc(migracao_rejeitada.id)).limit(100),
    candidatos: await tx.select().from(stg_trello_cartao).where(isNotNull(stg_trello_cartao.parecer_candidato)),
  });
});

/** Os cinco relatórios da Fase 4 — expõem, nunca resolvem sozinhos. */
rotas.get("/importar/reconciliacao", async (c) => {
  const usuario = await usuarioCom(c, "catalogo.gerenciar");
  const tx = c.get("tx");
  return pagina(c, "paginas/reconciliacao.html", usuario, {
    reconciliacao: await servico_reconciliacao.reconciliar(tx),
    rejeitadas: await servico_reconciliacao.rejeitadas(tx),
  });
});

rotas.get("/importar/reconciliacao.csv", async (c) => {
  await usuarioCom(c, "exportar");
  const resultado = await servico_reconciliacao.reconciliar(c.get("tx"));
  let texto = linha_csv(["Relatório", "Referência", "Detalhe"]);
  for (const [titulo, , achados] of resultado.como_secoes()) {
    for (const achado of achados) texto += linha_csv([titulo, achado.referencia, achado.detalhe]);
  }
  marcarDownload(c);
  return c.body("﻿" + texto, 200, {
    "content-type": "text/csv; charset=utf-8",
    "content-disposition": 'attachment; filename="reconciliacao.csv"',
  });
});

/** O arquivo do formulário, com a cópia guardada em `entrada/` do armazenamento. */
async function _receber(c: Ctx, padrao: string) {
  const f = await formulario(c);
  const arquivo = f.dados.get("arquivo");
  if (!(arquivo instanceof File)) throw new ErroHttp(422, "Campo obrigatório ausente: arquivo");
  const nome = (arquivo.name || padrao).replace(/[/\\\x00-\x1f]/g, "_") || padrao;
  const conteudo = new Uint8Array(await arquivo.arrayBuffer());
  await obterArmazenamento().gravar(`entrada/${nome}`, conteudo, arquivo.type || "application/octet-stream");
  return { f, nome, conteudo };
}

rotas.post("/importar/planilha", async (c) => {
  const usuario = await usuarioCom(c, "catalogo.gerenciar");
  const tx = c.get("tx");
  const { f, nome, conteudo } = await _receber(c, "planilha.xlsx");
  const aplicar = f.texto("aplicar") === "1";
  const relatorio = await importacao_planilha.importar(tx, { nome, conteudo }, usuario, aplicar);
  if (aplicar) await recontar_sequencia(tx);
  else desfazer(c);
  return pagina(c, "paginas/importar_relatorio.html", usuario, { relatorio, aplicado: aplicar, origem: "planilha" });
});

rotas.post("/importar/trello", async (c) => {
  const usuario = await usuarioCom(c, "catalogo.gerenciar");
  const tx = c.get("tx");
  const { f, conteudo } = await _receber(c, "trello.json");
  const aplicar = f.texto("aplicar") === "1";
  let dados: Record<string, unknown>;
  try {
    dados = JSON.parse(new TextDecoder().decode(conteudo));
  } catch {
    throw new ErroHttp(422, "O arquivo não é um JSON válido do quadro do Trello.");
  }
  const relatorio = await importacao_trello.importar(
    tx,
    dados,
    usuario,
    aplicar,
    // sem marcar, nada é baixado: os anexos entram no relatório de perdidos
    f.texto("baixar_anexos") === "1" ? importacao_trello.baixar_url : null,
  );
  if (!aplicar) desfazer(c);
  return pagina(c, "paginas/importar_relatorio_trello.html", usuario, {
    // o conjunto vira lista ordenada: o Nunjucks não percorre `Set`
    relatorio: Object.assign(relatorio, { listas_desconhecidas: [...relatorio.listas_desconhecidas].sort() }),
    aplicado: aplicar,
  });
});
